Visualization/generate_BackBLAST_heatmap.R

```diff
@@ -0,0 +1,618 @@
+#' To be mapped onto the main tree during plotting.
+#' @export
+generate_bootstrap_labels <- function(phylo_tree, bootstrap_cutoff) {
+  # For adding bootstrap labels like this, see https://guangchuangyu.github.io/software/ggtree/faq/ (accessed Sept 14, 2018)
+  # Note that 'labels' encompasses tip labels and node labels. You can to extract the tip labels alone.
+  
+  # Extract info from the tree into data frame format
+  bootstrap_label_data <- ggtree::ggtree(phylo_tree)$data
+  
+  # Set the labels to numeric (instead of character; will make non character go 'NA', like perhaps tip labels)
+  # This longer lapply function tries to avoid warnings from coersion while still allowing for transparent error reporting otherwise
+  bootstrap_label_data$label <- unlist(lapply(bootstrap_label_data$label, function(x) {
+    if(suppressWarnings(as.numeric(x)) %>% is.na == FALSE) {
+      return(as.numeric(x))
+    } else {
+      return(NA)}
+  }))
+  
+  # Filter out tip labels and 'NA' labels
+  bootstrap_label_data <- dplyr::filter(bootstrap_label_data, is.na(label) == FALSE & isTip == FALSE)
+  
+  # Set bootstrap cutoff if desired
+  if ( is.na(bootstrap_cutoff) == FALSE ) {
+    futile.logger::flog.info(glue::glue("Applying bootstrap display cutoff of ", bootstrap_cutoff))
+    bootstrap_label_data <- dplyr::filter(bootstrap_label_data, label >= bootstrap_cutoff)
+  }
+  
+  # Return whole tree data frame (helps for mapping the labels onto the existing tree later)
+  return(bootstrap_label_data)
+  
+}
+
+
+#' Makes an initial plot of the tree with overlaid bootstrap labels
+#' 
+#' @param phylo_tree ggtree-format tree
+#' @param bootstrap_label_data data frame extracted from tree and possibly modified, generated in 'generate_bootstrap_labels'
+#' @return ggtree plot
+#' @export
+plot_ggtree <- function(phylo_tree, bootstrap_label_data) {
+  
+  tree_plot <- ggtree::ggtree(phylo_tree, size = 1, colour = "black", ladderize = TRUE,
+                      branch.length = 0.1) +
+    
+    # Add the bootstrap labels from the external file
+    geom_text(data = bootstrap_label_data, aes(label = label), nudge_x = -0.035, nudge_y = 0.35, size = 3) +
+    
+    # Add the dotted lines from the tree tips
+    geom_tiplab(align = TRUE, linetype = "dotted", size = 0.1, offset = 0.1) +
+    
+    # Add scale bar
+    geom_treescale(x = 0, y = 8, linesize = 1, fontsize = 3, offset = 0.5) +
+    
+    # Manually tune the y-axis boundaries to match the heatmap: https://stackoverflow.com/a/18718196 (accessed Sept. 15, 2018)
+    # TODO - this might need to be a function of the number of entries
+    scale_y_discrete(expand = c(0,0.6)) +
+  
+    ## Manually set the righthand cutoff for the tree
+    xlim_tree(1)
+    # xlim etc.: https://guangchuangyu.github.io/2016/10/xlim_tree-set-x-axis-limits-for-only-tree-panel/ (accessed Sept 15, 2018)
+  
+  return(tree_plot)
+  
```

Comment:
Remove spaces below the final return. Change here and throughout.

Reply:
Done